### PR TITLE
Singleton Codecs

### LINKED ISSUE #196: Enhance Codec Implementation to Support Singleton Pattern
# Background
In the current state of the project, the Codec implementation is not designed as a singleton, which means that each time a new instance is required, a new object is created. This approach may lead to increased memory usage and unnecessary object creation overhead, especially in situations where a single instance of the Codec is sufficient for the entire application.

# Proposal
To address this issue, it is recommended to modify the Codec implementation to follow the singleton pattern. This would ensure that only one instance of the Codec exists throughout the application's lifecycle, providing benefits such as reduced memory consumption and improved performance.

### PR DESCRIPTION
**Pull request by jchrys**
Singleton Codecs

Motivation:
A single instance of the Codec is sufficient for the entire application.

Modification:
Implement Singleton pattern for codecs.

Result:
Reduced memory consumption and improved performance.
resolves #196

### CODE REVIEW

File: src/main/java/io/asyncer/r2dbc/mysql/MySqlConnection.java [hunk 1]
```diff
@@ -307,7 +307,7 @@ public IsolationLevel getTransactionIsolationLevel() {
     public Mono<Void> setTransactionIsolationLevel(IsolationLevel isolationLevel) {
         requireNonNull(isolationLevel, "isolationLevel must not be null");
 
-        // Set next transaction isolation level.
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/MySqlConnection.java]: Restore this line

File: src/main/java/io/asyncer/r2dbc/mysql/MySqlParameter.java [hunk 2]
```diff
@@ -52,7 +53,7 @@ default boolean isNull() {
      *
      * @return the encoded binary buffer(s).
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/MySqlParameter.java:56]: modify javadoc

File: src/main/java/io/asyncer/r2dbc/mysql/codec/BooleanCodec.java [hunk 3]
```diff
@@ -29,8 +29,10 @@
  */
 final class BooleanCodec extends AbstractPrimitiveCodec<Boolean> {
 
-    BooleanCodec(ByteBufAllocator allocator) {
-        super(allocator, Boolean.TYPE, Boolean.class);
+    public static final BooleanCodec INSTANCE = new BooleanCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/BooleanCodec.java]: package private

File: src/main/java/io/asyncer/r2dbc/mysql/codec/ByteCodec.java [hunk 4]
```diff
@@ -29,8 +29,10 @@
  */
 final class ByteCodec extends AbstractPrimitiveCodec<Byte> {
 
-    ByteCodec(ByteBufAllocator allocator) {
-        super(allocator, Byte.TYPE, Byte.class);
+    public static final ByteCodec INSTANCE = new ByteCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/ByteCodec.java]: package private

File: src/main/java/io/asyncer/r2dbc/mysql/codec/DoubleCodec.java [hunk 5]
```diff
@@ -31,8 +31,10 @@
  */
 final class DoubleCodec extends AbstractPrimitiveCodec<Double> {
 
-    DoubleCodec(ByteBufAllocator allocator) {
-        super(allocator, Double.TYPE, Double.class);
+    public static final DoubleCodec INSTANCE = new DoubleCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/DoubleCodec.java]: package private

File: src/main/java/io/asyncer/r2dbc/mysql/codec/FloatCodec.java [hunk 6]
```diff
@@ -31,8 +31,10 @@
  */
 final class FloatCodec extends AbstractPrimitiveCodec<Float> {
 
-    FloatCodec(ByteBufAllocator allocator) {
-        super(allocator, Float.TYPE, Float.class);
+    public static final FloatCodec INSTANCE = new FloatCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/FloatCodec.java]: package private

File: src/main/java/io/asyncer/r2dbc/mysql/codec/IntegerCodec.java [hunk 7]
```diff
@@ -34,8 +34,10 @@
  */
 final class IntegerCodec extends AbstractPrimitiveCodec<Integer> {
 
-    IntegerCodec(ByteBufAllocator allocator) {
-        super(allocator, Integer.TYPE, Integer.class);
+    public static final IntegerCodec INSTANCE = new IntegerCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/IntegerCodec.java]: package private

File: src/main/java/io/asyncer/r2dbc/mysql/codec/LongCodec.java [hunk 8]
```diff
@@ -35,8 +35,10 @@
  */
 final class LongCodec extends AbstractPrimitiveCodec<Long> {
 
-    LongCodec(ByteBufAllocator allocator) {
-        super(allocator, Long.TYPE, Long.class);
+    public static final LongCodec INSTANCE = new LongCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/LongCodec.java]: package private

File: src/main/java/io/asyncer/r2dbc/mysql/codec/ShortCodec.java [hunk 9]
```diff
@@ -30,8 +30,10 @@
  */
 final class ShortCodec extends AbstractPrimitiveCodec<Short> {
 
-    ShortCodec(ByteBufAllocator allocator) {
-        super(allocator, Short.TYPE, Short.class);
+    public static final ShortCodec INSTANCE = new ShortCodec();
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/codec/ShortCodec.java]: package private

**Review by mirromutth — APPROVED** [review]
LGTM

**Pull request merged**